src/te_checks.c

```diff
@@ -140,6 +140,32 @@ struct check_result *check_unordered_perms(__attribute__((unused)) const struct
 
 }
 
+struct check_result *check_no_self(__attribute__((unused)) const struct check_data *data,
+                                   const struct policy_node *node)
+{
+	if (node->flavor != NODE_AV_RULE && node->flavor != NODE_XAV_RULE) {
+		return alloc_internal_error("Bad node type given to check C-007");
+	}
+	struct av_rule_data *av_data = node->data.av_data;
+
+	if (av_data->sources->next ||
+	    av_data->targets->next ||
+	    0 == strcmp(av_data->targets->string, "self") ||
+	    0 != strcmp(av_data->sources->string, av_data->targets->string)) {
+		return NULL;
+	}
+
+	if (av_data->sources->string[0] != '$' &&
+	    !ends_with(av_data->sources->string, strlen(av_data->sources->string), "_t", 2) &&
+	     !look_up_in_decl_map(av_data->sources->string, DECL_TYPE)) {
+		return NULL;
+	}
```

Comment:
This will not skip `allow $1 $1:fd use;`

Maybe use
```c
if ((av_data->sources->string[0] != '$' ||
    !ends_with(av_data->sources->string, strlen(av_data->sources->string), "_t", 2)) &&
     !look_up_in_decl_map(av_data->sources->string, DECL_TYPE)) {
	return NULL;
}
```
?

Reply:
I'm getting overly confused with the complicated boolean expressions, so I've just gone with nested if/else for clarity.  Hopefully it's not just right now, it's also clearly right.